apis/python/tests/test_dataframe.py

```diff
@@ -461,6 +468,229 @@ def test_bool_enums(tmp_path, data_and_expected_levels):
         assert actual == expect
 
 
+@pytest.mark.parametrize("extend_not_write", [False, True])
+def test_extend_enumeration_values(tmp_path, extend_not_write):
+    """Compares the older create+write path against create+extend path"""
+    uri = tmp_path.as_posix()
+    domain = [[0, 2]]
+
+    schema = pa.schema(
+        {
+            "soma_joinid": pa.int64(),
+            "not_an_enum": pa.large_string(),
+            "int64_enum1": pa.dictionary(pa.int8(), pa.int64()),
+            "int64_enum2": pa.dictionary(pa.int8(), pa.int64()),
+            "int64_enum3": pa.dictionary(pa.int8(), pa.int64()),
+            "float64_enum": pa.dictionary(pa.int8(), pa.float64()),
+            "string_enum1": pa.dictionary(pa.int32(), pa.large_string()),
+            "string_enum2": pa.dictionary(pa.int32(), pa.large_string()),
+            "bool_enum1": pa.dictionary(pa.int32(), pa.bool_()),
+            "bool_enum2": pa.dictionary(pa.int32(), pa.bool_()),
+        }
+    )
+    enum_column_names = [
+        name for name in schema.names if pa.types.is_dictionary(schema.field(name).type)
+    ]
+
+    pandas_data = {
+        "soma_joinid": [0, 1, 2],
+        "not_an_enum": ["quick", "brown", "fox"],
+        "int64_enum1": pd.Categorical([55555, 55555, 7777777]),
+        "int64_enum2": pd.Categorical(np.array([55555, 55555, 7777777])),
+        "int64_enum3": pd.Categorical(
+            np.array([7777777, 55555, 55555], dtype=np.int64)
+        ),
+        "float64_enum": pd.Categorical(np.array([2.5, 8.875, 2.5], dtype=np.float64)),
+        "string_enum1": pd.Categorical(["hello", "hello", "goodbye"]),
+        "string_enum2": pd.Categorical(["goodbye", "goodbye", "hello"]),
+        "bool_enum1": pd.Categorical([True, True, False]),
+        "bool_enum2": pd.Categorical([False, False, True]),
+    }
+    arrow_data = pa.Table.from_pydict(pandas_data)
+
+    expected_levels = {
+        "int64_enum1": [55555, 7777777],
+        "int64_enum2": [55555, 7777777],
+        "int64_enum3": [55555, 7777777],
+        "float64_enum": [2.5, 8.875],
+        "string_enum1": ["goodbye", "hello"],
+        "string_enum2": ["goodbye", "hello"],
+        "bool_enum1": [False, True],
+        "bool_enum2": [False, True],
+    }
+    arrow_data = pa.Table.from_pydict(pandas_data)
+
+    # Note: doing the .categories bit is implicitly deduplicating the data
+    # before it's sent to tiledbsoma. Other tests, not this one, stress the
+    # deduplicate-or-not logic.
+    values = {
+        name: pa.array(pandas_data[name].categories) for name in enum_column_names
+    }
+
+    # Do the extend without write, or write with implicit extend
+    with soma.DataFrame.create(
+        uri,
+        schema=schema,
+        domain=domain,
+        index_column_names=["soma_joinid"],
+    ) as sdf:
+        if extend_not_write:
```

Comment:
this is missing another common case where:
* the dataframe has been written to (e.g., enum values for the column are set to something)
* then the enum is extended

In other words, you current test extension on an empty dataframe, but not on one that already has data

Reply:
@bkmartinjr did you miss seeing

* https://github.com/single-cell-data/TileDB-SOMA/blob/b3965b35265540b61d4bbade6fd247ebc0eb7dc6/apis/python/tests/test_dataframe.py#L472
* https://github.com/single-cell-data/TileDB-SOMA/blob/b3965b35265540b61d4bbade6fd247ebc0eb7dc6/apis/python/tests/test_dataframe.py#L586

I intentionally created more than one unit-test functio; testing an extension on a dataframe that already has data is in the second one

Reply:
I'll add a comment in the first one that mentions there is another test case after it

Reply:
With the latest changes, I still don't see a test for the sequence I mentioned above:
1. create
2. write (with implicit extend)
3. extend

The second test case (`test-extend_enumeration_values_deduplication`) does not perform any `write()` operations.  And the first test case does a `write()`, but not followed by an `extend`.

I realize this is a very niche edge case -- I'm just trying to make sure that when we use both `write()` and `extend_enumeration()` in combination, the work as expected

Reply:
https://github.com/single-cell-data/TileDB-SOMA/pull/3815/commits/979e4c1209121fbc026a2f10ae22645e9d37d65c

Reply:
pushed